File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/util/ErrorMessageFormatterTest.java
package com.fincity.nocode.kirun.engine.util;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.json.schema.array.ArraySchemaType;
import com.fincity.nocode.kirun.engine.json.schema.type.SchemaType;
import com.google.gson.JsonPrimitive;

class ErrorMessageFormatterTest {

    @Test
    void testFormatValue_Null() {
        assertEquals("null", ErrorMessageFormatter.formatValue(null));
    }

    @Test
    void testFormatValue_Strings() {
        assertEquals("\"hello\"", ErrorMessageFormatter.formatValue("hello"));
    }

    @Test
    void testFormatValue_Numbers() {
        assertEquals("42", ErrorMessageFormatter.formatValue(42));
        assertEquals("3.14", ErrorMessageFormatter.formatValue(3.14));
    }

    @Test
    void testFormatValue_Booleans() {
        assertEquals("true", ErrorMessageFormatter.formatValue(true));
        assertEquals("false", ErrorMessageFormatter.formatValue(false));
    }

    @Test
    void testFormatValue_Objects() {
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", "John");
        obj.put("age", 30);

        String result = ErrorMessageFormatter.formatValue(obj);
        assertTrue(result.contains("\"name\""));
        assertTrue(result.contains("\"John\""));
        assertTrue(result.contains("\"age\""));
        assertTrue(result.contains("30"));
    }

    @Test
    void testFormatValue_Arrays() {
        List<Integer> arr = List.of(1, 2, 3);
        String result = ErrorMessageFormatter.formatValue(arr);

        assertTrue(result.contains("["));
        assertTrue(result.contains("1"));
        assertTrue(result.contains("2"));
        assertTrue(result.contains("3"));
        assertTrue(result.contains("]"));
    }

    @Test
    void testFormatValue_CircularReferences() {
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", "test");
        obj.put("self", obj);  // Circular reference

        String result = ErrorMessageFormatter.formatValue(obj);
        assertTrue(result.contains("[Circular]"));
    }

    @Test
    void testFormatValue_Truncation() {
        Map<String, String> longObj = new HashMap<>();
        longObj.put("data", "x".repeat(300));

        String result = ErrorMessageFormatter.formatValue(longObj, 50);
        assertTrue(result.length() <= 53); // 50 + "..."
        assertTrue(result.endsWith("..."));
    }

    @Test
    void testFormatFunctionName_WithNamespace() {
        assertEquals("UIEngine.SetStore",
                ErrorMessageFormatter.formatFunctionName("UIEngine", "SetStore"));
    }

    @Test
    void testFormatFunctionName_WithoutNamespace() {
        assertEquals("loadEverything",
                ErrorMessageFormatter.formatFunctionName(null, "loadEverything"));
    }

    @Test
    void testFormatFunctionName_EmptyNamespace() {
        assertEquals("loadApp",
                ErrorMessageFormatter.formatFunctionName("", "loadApp"));
    }

    @Test
    void testFormatFunctionName_UndefinedString() {
        assertEquals("loadEverything",
                ErrorMessageFormatter.formatFunctionName("undefined", "loadEverything"));
    }

    @Test
    void testFormatFunctionName_NullString() {
        assertEquals("loadEverything",
                ErrorMessageFormatter.formatFunctionName("null", "loadEverything"));
    }

    @Test
    void testFormatStatementName_Valid() {
        assertEquals("'storeString'",
                ErrorMessageFormatter.formatStatementName("storeString"));
    }

    @Test
    void testFormatStatementName_Null() {
        assertNull(ErrorMessageFormatter.formatStatementName(null));
    }

    @Test
    void testFormatStatementName_Empty() {
        assertNull(ErrorMessageFormatter.formatStatementName(""));
    }

    @Test
    void testFormatStatementName_UndefinedString() {
        assertNull(ErrorMessageFormatter.formatStatementName("undefined"));
    }

    @Test
    void testFormatSchemaDefinition_SingleType() {
        Schema schema = Schema.ofInteger("test");
        assertEquals("Integer", ErrorMessageFormatter.formatSchemaDefinition(schema));
    }

    @Test
    void testFormatSchemaDefinition_ArrayType() {
        ArraySchemaType ast = new ArraySchemaType();
        ast.setSingleSchema(Schema.ofInteger("item"));

        Schema schema = Schema.ofArray("test").setItems(ast);
        assertEquals("Array<Integer>", ErrorMessageFormatter.formatSchemaDefinition(schema));
    }

    @Test
    void testFormatSchemaDefinition_TupleType() {
        ArraySchemaType ast = new ArraySchemaType();
        ast.setTupleSchema(List.of(
            Schema.ofString("first"),
            Schema.ofInteger("second")
        ));

        Schema schema = Schema.ofArray("test").setItems(ast);
        assertEquals("[String, Integer]", ErrorMessageFormatter.formatSchemaDefinition(schema));
    }

    @Test
    void testFormatSchemaDefinition_EnumType() {
        Schema schema = Schema.ofString("test").setEnums(List.of(
            new JsonPrimitive("option1"),
            new JsonPrimitive("option2")
        ));

        String result = ErrorMessageFormatter.formatSchemaDefinition(schema);
        assertTrue(result.contains("String"));
        assertTrue(result.contains("option1"));
        assertTrue(result.contains("option2"));
    }

    @Test
    void testFormatSchemaDefinition_MultipleTypes() {
        Schema schema = Schema.of("test", SchemaType.STRING, SchemaType.INTEGER);
        String result = ErrorMessageFormatter.formatSchemaDefinition(schema);

        assertTrue(result.contains("String") || result.contains("Integer"));
        assertTrue(result.contains("|"));
    }

    @Test
    void testBuildFunctionExecutionError_WithStatementName() {
        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "UIEngine.SetStore",
            "'storeString'",
            "Expected an array but found {\"key\": \"value\"}"
        );

        assertEquals("Error while executing the function UIEngine.SetStore in statement 'storeString': "
            + "Expected an array but found {\"key\": \"value\"}", result);
    }

    @Test
    void testBuildFunctionExecutionError_WithoutStatementName() {
        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "loadEverything",
            null,
            "Some error occurred"
        );

        assertEquals("Error while executing the function loadEverything: Some error occurred", result);
    }

    @Test
    void testBuildFunctionExecutionError_WithParameterName() {
        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "UIEngine.SetStore",
            "'storeString'",
            "Expected an array but found {}",
            "value"
        );

        assertEquals("Error while executing the function UIEngine.SetStore's parameter value in statement 'storeString': "
            + "Expected an array but found {}", result);
    }

    @Test
    void testBuildFunctionExecutionError_WithParameterNoStatement() {
        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "loadApp",
            null,
            "Invalid parameter",
            "config"
        );

        assertEquals("Error while executing the function loadApp's parameter config: Invalid parameter", result);
    }

    @Test
    void testBuildFunctionExecutionError_WithParameterSchema() {
        ArraySchemaType ast = new ArraySchemaType();
        ast.setSingleSchema(Schema.ofInteger("item"));
        Schema schema = Schema.ofArray("value").setItems(ast);

        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "System.Array.Concatenate",
            null,
            "Expected an array but found null",
            "secondSource",
            schema
        );

        assertEquals("Error while executing the function System.Array.Concatenate's parameter secondSource "
            + "[Expected: Array<Integer>]: Expected an array but found null", result);
    }

    @Test
    void testBuildFunctionExecutionError_NestedErrors() {
        String innerError = "Error while executing the function UIEngine.SetStore in statement 'storeString' "
            + "[Expected: Array]: Expected an array but found {}";

        String result = ErrorMessageFormatter.buildFunctionExecutionError(
            "loadApp",
            null,
            innerError
        );

        assertEquals("Error while executing the function loadApp: \n" + innerError, result);
    }

    @Test
    void testFormatErrorMessage_ThrowableWithMessage() {
        Throwable error = new RuntimeException("Something went wrong");
        assertEquals("Something went wrong", ErrorMessageFormatter.formatErrorMessage(error));
    }

    @Test
    void testFormatErrorMessage_String() {
        assertEquals("Error message", ErrorMessageFormatter.formatErrorMessage("Error message"));
    }

    @Test
    void testFormatErrorMessage_Null() {
        assertEquals("Unknown error", ErrorMessageFormatter.formatErrorMessage((Throwable) null));
    }

    @Test
    void testFormatErrorMessage_Object() {
        Map<String, String> error = Map.of("message", "Expected an array but found something");
        String result = ErrorMessageFormatter.formatErrorMessage(error);

        assertTrue(result.contains("message"));
        assertTrue(result.contains("Expected an array but found something"));
    }
}

File: kirun-js/__tests__/engine/util/ErrorMessageFormatterTest.ts
import { ErrorMessageFormatter } from '../../../src/engine/util/ErrorMessageFormatter';
import { Schema, ArraySchemaType } from '../../../src';

describe('ErrorMessageFormatter', () => {
    describe('formatValue', () => {
        test('should format null', () => {
            expect(ErrorMessageFormatter.formatValue(null)).toBe('null');
        });

        test('should format undefined', () => {
            expect(ErrorMessageFormatter.formatValue(undefined)).toBe('undefined');
        });

        test('should format strings with quotes', () => {
            expect(ErrorMessageFormatter.formatValue('hello')).toBe('"hello"');
        });

        test('should format numbers', () => {
            expect(ErrorMessageFormatter.formatValue(42)).toBe('42');
            expect(ErrorMessageFormatter.formatValue(3.14)).toBe('3.14');
        });

        test('should format booleans', () => {
            expect(ErrorMessageFormatter.formatValue(true)).toBe('true');
            expect(ErrorMessageFormatter.formatValue(false)).toBe('false');
        });

        test('should format objects as JSON', () => {
            const obj = { name: 'John', age: 30 };
            const result = ErrorMessageFormatter.formatValue(obj);
            expect(result).toContain('"name"');
            expect(result).toContain('"John"');
            expect(result).toContain('"age"');
            expect(result).toContain('30');
        });

        test('should format arrays as JSON', () => {
            const arr = [1, 2, 3];
            const result = ErrorMessageFormatter.formatValue(arr);
            expect(result).toContain('[');
            expect(result).toContain('1');
            expect(result).toContain('2');
            expect(result).toContain('3');
            expect(result).toContain(']');
        });

        test('should handle circular references', () => {
            const obj: any = { name: 'test' };
            obj.self = obj;
            const result = ErrorMessageFormatter.formatValue(obj);
            expect(result).toContain('[Circular]');
        });

        test('should truncate long values', () => {
            const longObj = { data: 'x'.repeat(300) };
            const result = ErrorMessageFormatter.formatValue(longObj, 50);
            expect(result.length).toBeLessThanOrEqual(53); // 50 + "..."
            expect(result).toContain('...');
        });
    });

    describe('formatFunctionName', () => {
        test('should format namespace.name', () => {
            expect(ErrorMessageFormatter.formatFunctionName('UIEngine', 'SetStore')).toBe(
                'UIEngine.SetStore',
            );
        });

        test('should handle undefined namespace', () => {
            expect(ErrorMessageFormatter.formatFunctionName(undefined, 'loadEverything')).toBe(
                'loadEverything',
            );
        });

        test('should handle null namespace', () => {
            expect(ErrorMessageFormatter.formatFunctionName(null, 'loadApp')).toBe('loadApp');
        });

        test('should handle "undefined" string', () => {
            expect(ErrorMessageFormatter.formatFunctionName('undefined', 'loadEverything')).toBe(
                'loadEverything',
            );
        });
    });

    describe('formatStatementName', () => {
        test('should format statement name with quotes', () => {
            expect(ErrorMessageFormatter.formatStatementName('storeString')).toBe("'storeString'");
        });

        test('should return null for undefined statement name', () => {
            expect(ErrorMessageFormatter.formatStatementName(undefined)).toBeNull();
        });

        test('should return null for null statement name', () => {
            expect(ErrorMessageFormatter.formatStatementName(null)).toBeNull();
        });

        test('should return null for "undefined" string', () => {
            expect(ErrorMessageFormatter.formatStatementName('undefined')).toBeNull();
        });
    });

    describe('buildFunctionExecutionError', () => {
        test('should build error message with statement name', () => {
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'UIEngine.SetStore',
                "'storeString'",
                'Expected an array but found {"key": "value"}',
            );
            expect(result).toBe(
                "Error while executing the function UIEngine.SetStore in statement 'storeString': Expected an array but found {\"key\": \"value\"}",
            );
        });

        test('should build error message without statement name', () => {
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'loadEverything',
                null,
                'Some error occurred',
            );
            expect(result).toBe(
                'Error while executing the function loadEverything: Some error occurred',
            );
        });

        test('should build error message with parameter name', () => {
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'UIEngine.SetStore',
                "'storeString'",
                'Expected an array but found {}',
                'value',
            );
            expect(result).toBe(
                "Error while executing the function UIEngine.SetStore's parameter value in statement 'storeString': Expected an array but found {}",
            );
        });

        test('should build error message with parameter but no statement name', () => {
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'loadApp',
                null,
                'Invalid parameter',
                'config',
            );
            expect(result).toBe(
                "Error while executing the function loadApp's parameter config: Invalid parameter",
            );
        });

        test('should build error message with parameter schema definition', () => {
            const ast = new ArraySchemaType();
            ast.setSingleSchema(Schema.ofInteger('item'));
            const schema = Schema.ofArray('value').setItems(ast);
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'System.Array.Concatenate',
                null,
                'Expected an array but found null',
                'secondSource',
                schema,
            );
            expect(result).toBe(
                "Error while executing the function System.Array.Concatenate's parameter secondSource [Expected: Array<Integer>]: Expected an array but found null",
            );
        });

        test('should wrap nested error messages to show execution path', () => {
            // Simulate nested function calls where inner error is already formatted
            const innerError = "Error while executing the function UIEngine.SetStore in statement 'storeString' [Expected: Array]: Expected an array but found {}";

            // Wrap it from an outer function to show call chain
            const result = ErrorMessageFormatter.buildFunctionExecutionError(
                'loadApp',
                null,
                innerError,
            );

            // Should wrap the error with a newline to show the full execution path
            expect(result).toBe("Error while executing the function loadApp: \n" + innerError);
        });
    });

    describe('formatErrorMessage', () => {
        test('should extract message from Error object', () => {
            const error = new Error('Something went wrong');
            expect(ErrorMessageFormatter.formatErrorMessage(error)).toBe('Something went wrong');
        });

        test('should handle string errors', () => {
            expect(ErrorMessageFormatter.formatErrorMessage('Error message')).toBe('Error message');
        });

        test('should handle undefined error', () => {
            expect(ErrorMessageFormatter.formatErrorMessage(undefined)).toBe('Unknown error');
        });

        test('should handle object errors', () => {
            const error = { message: 'Expected an array but found something' };
            const result = ErrorMessageFormatter.formatErrorMessage(error);
            expect(result).toBe('Expected an array but found something');
        });
    });
});
